agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java

```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
       }
 
       private void detectDocker(@NotNull final BuildAgentConfiguration config) {
-        if (!SystemInfo.isLinux) {
-          LOG.debug("Docker is only available under Linux");
-          return;
-        }
 
-        final String output = executeCommandWithShell("docker", "docker --version");
+        final String[] output = executeCommandWithShell("docker", "docker version");
         if (output == null) return;
 
-        String ver = output.toLowerCase().trim();
-        ver = ver.replaceAll("\\s*docker\\s+version\\s+", "");
-        ver = ver.replaceAll(",?\\s+build\\s+", "-");
+        String[] cleanedArray = new String[output.length];
 
-        if (StringUtil.isEmptyOrSpaces(ver)) {
-          LOG.warn("Failed to parse docker version: " + output);
+        for (int i = 0; i < output.length; i++)
+          cleanedArray[i] = output[i].trim();
+
+        int serverPosition = Arrays.asList(cleanedArray).indexOf("Server:");
+        String[] clientProperties = Arrays.copyOfRange(cleanedArray, 0, serverPosition - 1);
+        String[] serverProperties = Arrays.copyOfRange(cleanedArray, serverPosition, cleanedArray.length);
+
+        String clientVersion = getDockerPropertyValue(clientProperties, "Version:");
+        String clientOsArch = getDockerPropertyValue(clientProperties, "OS/Arch:");;
+        String serverVersion = getDockerPropertyValue(serverProperties, "Version:");;
+        String serverOsArch = getDockerPropertyValue(serverProperties, "OS/Arch:");;
+
+        if (StringUtil.isEmptyOrSpaces(clientVersion)
+                || StringUtil.isEmptyOrSpaces(clientOsArch)
+                || StringUtil.isEmptyOrSpaces(serverVersion)
+                || StringUtil.isEmptyOrSpaces(serverOsArch)) {
+          LOG.warn("Failed to parse docker information: " + Arrays.toString(output));
           return;
         }
 
-        config.addConfigurationParameter(VMConstants.DOCKER_PROPERTY, ver);
+        config.addConfigurationParameter(VMConstants.DOCKER_PROPERTY, clientVersion);
+        config.addConfigurationParameter(VMConstants.DOCKER_CLIENT_VERSION_PROPERTY, clientVersion);
+        config.addConfigurationParameter(VMConstants.DOCKER_CLIENT_OS_ARCH_PROPERTY, clientOsArch);
+        config.addConfigurationParameter(VMConstants.DOCKER_SERVER_VERSION_PROPERTY, serverVersion);
+        config.addConfigurationParameter(VMConstants.DOCKER_SERVER_OS_ARCH_PROPERTY, serverOsArch);
+
+      }
+
+      @Nullable
+      private String getDockerPropertyValue(String[] dockerProperties, String key) {
+        for (String line : dockerProperties) {
+          if (line.contains(key)) {
+            return line.replace(key, "").trim();
+          }
+        }
+        return null;
       }
 
 
       @Nullable
-      private String executeCommandWithShell(@NotNull final String name, @NotNull final String... command) {
+      private String[] executeCommandWithShell(@NotNull final String name, @NotNull final String... command) {
```

Comment:
It's better to return `String` as it used to be and process the output carefully around docker detection